### PR TITLE
fix(sqlite): Prevent reaching the maximum number of parameters in an SQL statement

### PR DESCRIPTION
**Pull request by Hywan**
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/2224.

This patch introduces a new `chunk_large_query_over` function. Imagine there is a _dynamic_ query that runs potentially large number of parameters, so much that the maximum of parameters can be hit. Then, this function is for you. It will execute the query on chunks of parameters.

`chunk_large_query_over` uses `Vec::split_off` to avoid cloning `Key`s as much as possible. It's difficult to use references here because of the async nature of `SqliteObjectExt::prepare`.

This patch updates `get_kv_blobs`, `get_room_infos`, `get_maybe_stripped_state_events_for_keys`, `get_profiles`, `get_user_ids`, and `get_display_names` to use this new function.

This patch finally adds the `repeat_vars` function to replace in a more efficient way the `vec!["?"; n].join(", ")` pattern. There is less memory allocations.

Note: I've added an `assert_ne!(count, 0);` in `repeat_vars` but I don't like that. Ideally, we should be sure that the `Vec<Key>` used to create the chunk cannot be empty. I consider this is out of the scope of this PR and can be addressed later.

### CODE REVIEW

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 1]
```diff
@@ -594,20 +604,22 @@ trait SqliteObjectStateStoreExt: SqliteObjectExt {
         event_type: Key,
         state_keys: Vec<Key>,
     ) -> Result<Vec<(bool, Vec<u8>)>> {
-        let sql_params = vec!["?"; state_keys.len()].join(", ");
-        let sql = format!(
-            "SELECT stripped, data FROM state_event
+        chunk_large_query_over(state_keys, None, move |state_keys: Vec<Key>| {
+            let sql_params = repeat_vars(state_keys.len());
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:608]: Even more efficient:

```suggestion
            use itertools::Itertools;
            let sql_params = std::iter::repeat("?").take(state_keys.len()).format(", ");
```

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 2]
```diff
@@ -1531,6 +1551,55 @@ struct ReceiptData {
...
+    };
+
+    // `Limit` has a `repr(i32)`, it's safe to cast it to `i32`. Then substract 1 in
+    // case of.
+    let maximum_chunk_size = Limit::SQLITE_LIMIT_VARIABLE_NUMBER as i32 - 1;
+    let maximum_chunk_size: usize = maximum_chunk_size
+        .try_into()
+        .map_err(|_| Error::SqliteMaximumVariableNumber(maximum_chunk_size))?;
+
+    while keys_to_chunk.is_empty().not() {
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:1581]:
```suggestion
    while !keys_to_chunk.is_empty() {
```

File: crates/matrix-sdk-sqlite/src/state_store.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk-sqlite/src/state_store.rs:608]: TIL, thanks!

**Review by jplatte — APPROVED** [review]

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 3]
```diff
@@ -1531,6 +1551,51 @@ struct ReceiptData {
...
+{
+    let mut all_results = if let Some(capacity) = result_capacity {
+        Vec::with_capacity(capacity)
+    } else {
+        Vec::new()
+    };
+
+    // `Limit` has a `repr(i32)`, it's safe to cast it to `i32`. Then subtract 1 in
+    // case of.
+    let maximum_chunk_size = Limit::SQLITE_LIMIT_VARIABLE_NUMBER as i32 - 1;
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:1576]: This won't work, will it? If the query contains a handful of non-chunked placeholders already, adding `SQLITE_LIMIT_VARIABLE_NUMBER - 1` placeholders will still overflow. I'd rather do `SQLITE_LIMIT_VARIABLE_NUMBER / 2` or sth. like that.

**Review by jplatte — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-sqlite/src/state_store.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk-sqlite/src/state_store.rs:1576]: Good catch!!

**Review by jplatte — APPROVED** [review]

**Pull request merged**